### PR TITLE
Optimize filter executor in pull-based executor

### PR DESCRIPTION
**Pull request by xudong963**
# Which issue does this PR close?

No

# Rationale for this change

If **Selection operator** doesn't produce any rows, we can just pull next `RecordBatch`es from it.

# What changes are included in this PR?

Add loop for filter executor.

# Are these changes tested?

Covered by existing tests

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by xudong963** [comment]
I think the more elegant way is to directly **Skip** current iterator to the next iterator.

For example:

```rust
trait A {
    fn execute() -> Result<Stream>;
}


Struct B {
    input: C,
}

Struct C {
    input: D
}

Struct D {
    
}

// All D, B, C have implemented trait A and trait Stream
// Each execute method will call its input's execute method


fn main {
    let b = B::new();
    let data_stream = b.execute();
    while let Some(stream) = data_stream.next().await {
        ...
    }
}

impl Stream for C {
    type Item = ..;

    fn poll_next(
        mut self: std::pin::Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Self::Item>> {
        ...
        if !predicate(value) {
            // Skip current iteration
        }
    }
}
```
But I don't find a proper way to implement it, `Stream` crate seems not to provide related API. Any thoughts? @tustvold

File: datafusion/core/src/physical_plan/filter.rs [hunk 1]
```diff
@@ -238,15 +238,31 @@ impl Stream for FilterExecStream {
...
-                timer.done();
-                Some(filtered_batch)
+        let poll;
+        loop {
+            match self.input.poll_next_unpin(cx) {
+                Poll::Ready(value) => match value {
+                    Some(Ok(batch)) => {
+                        let timer = self.baseline_metrics.elapsed_compute().timer();
+                        let filtered_batch = batch_filter(&batch, &self.predicate)?;
+                        if filtered_batch.num_rows() == 0 {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/filter.rs:248]:
```suggestion
                        // skip entirely filtered batches
                        if filtered_batch.num_rows() == 0 {
```

**Review by alamb — APPROVED** [review]
Seems like a reasonable change to me. Thanks @xudong963

**Review by Ted-Jiang — APPROVED** [review]
Nice find! 👍

**Comment by xudong963** [comment]
Thanks for your review!

**Pull request merged**